core/testcontainers/core/image.py

```diff
@@ -0,0 +1,70 @@
+from typing import TYPE_CHECKING, Optional
+
+from typing_extensions import Self
+
+from testcontainers.core.docker_client import DockerClient
+from testcontainers.core.utils import setup_logger
+
+if TYPE_CHECKING:
+    from docker.models.containers import Image
+
+logger = setup_logger(__name__)
+
+
+class DockerImage:
+    """
+    Basic image object to build Docker images.
+
+    .. doctest::
+            >>> from testcontainers.core.image import DockerImage
+            >>> with DockerImage(tag="new_image", path=".") as image:
+            ...    logs = image.get_logs()
+
+    :param tag: Tag of the image
+    :param path: Path to the Dockerfile
+    """
+
+    def __init__(
+        self,
+        tag: str,
+        path: str,
+        docker_client_kw: Optional[dict] = None,
+        **kwargs,
+    ) -> None:
+        self.tag = tag
+        self.path = path
+        self._docker = DockerClient(**(docker_client_kw or {}))
+        self._kwargs = kwargs
+
+    def build(self, **kwargs) -> Self:
+        logger.info(f"Building image {self.tag} from {self.path}")
+        docker_client = self.get_docker_client()
+        self._image, self._logs = docker_client.build(path=self.path, tag=self.tag, **kwargs)
+        logger.info(f"Built image {self.tag}")
+        return self
+
+    def remove(self, force=True, noprune=False) -> None:
+        """
+        Remove the image.
+
+        :param force: Remove the image even if it is in use
+        :param noprune: Do not delete untagged parent images
+        """
+        if self._image:
+            self._image.remove(force=force, noprune=noprune)
+        self.get_docker_client().client.close()
+
+    def __enter__(self) -> Self:
+        return self.build()
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        self.remove()
```

Comment:
rather than untagging - why not never tag and let it be cleaned up by `system prune`? im just thinking aloud - not 100% confident about how it works in java but the other pr for custom image also does what you did here, just challenging yall to think of something without the tagging/removal issue.

Reply:
Let me try and follow up on this notion (sorry for the wall of text)

let me start with saying I do think its possible, the tag in the image is not mandatory, and if kept as a ref to the image object we can probably use the id to be passed as the `image` to `DockerContainer`, this will have a very different workflow from what I presented in this commit as we don't really need a context manager. now having established this:

- as a concept I think tests should cleanup after themself 
-- for example in the case where we ran similar tests one after the other, in our case a very similar image with small changes and I want to make sure it does get rebuilt in the process, you can argue that sometimes we do want to reuse the image but from my experience a clean slate is preferable.
-- `system prune` is great but I tend to keep it for when I ran out of space or more generally when  something bad happened, as a user I don't want to run it after each test to keep a neat and clean env. (not saying it not possible ofc, just less preferable imo) 
- we can expanded this and draw similarities to the `DockerContainer` object.  as part of its workflow on `stop` it also does a targeted `remove`, so I can argue we are simply keeping the same behavior for but containers and images. or do we consider maybe not removing the container at the end and also letting the user clean it later? (I know this is kind of a stretch)
- regarding how `testcontainers-java` handle this, They use a more more elaborated system to handle images (for ref https://github.com/testcontainers/testcontainers-java/tree/d3b0df2d89ee349890c94b0ce220a167c82ddf51/core/src/main/java/org/testcontainers/images) I am not sure we can or should do exactly what they are doing (at least at this stage) I did invest some time digging the code (I'm no java expert) and they do have https://github.com/testcontainers/testcontainers-java/blob/d3b0df2d89ee349890c94b0ce220a167c82ddf51/core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java#L55 ```private boolean deleteOnExit = true;``` and checking with the official docs https://java.testcontainers.org/features/creating_images/#automatic-deletion confirms they do intend on removing the temp images, but I can't pinpoint the exact location they are doing it  (the only clue I have is that it happens on the JVM shutdown). 

TL;DR
I think the tagging can be changed to be optional (need to do some testing) and the removal is not an issue but a very important feature (implemented and included on the java version) which can be optional (defaults to true)